File: 백준/Silver/2468. 안전 영역/안전 영역.py
from collections import deque

import sys
input = sys.stdin.readline

def make_zero(h):
    for i in range(n):
        for j in range(n):
            if matrix[i][j] <= h:
                matrix[i][j] = 0

def is_reachable(x, y):
    return 0 <= x < n and 0 <= y < n and matrix[x][y] != 0

def bfs(x, y):
    dxs = [1, 0, -1, 0]
    dys = [0, 1, 0, -1]

    queue = deque([[x, y]])
    visited[x][y] = True

    while queue:
        x, y = queue.popleft()

        for dx, dy in zip(dxs, dys):
            nx, ny = x + dx, y + dy

            if is_reachable(nx, ny) and not visited[nx][ny]:
                visited[nx][ny] = True
                queue.append([nx, ny])

n = int(input())

max_height = 0
matrix = []
for _ in range(n):
    arr = list(map(int, input().split()))
    max_height = max(max_height, max(arr))
    matrix.append(arr)

max_cnt = 0

for h in range(max_height + 1):
    visited = [[False] * n for _ in range(n)]
    make_zero(h)

    cnt = 0
    for i in range(n):
        for j in range(n):
            if matrix[i][j] != 0 and not visited[i][j]:
                bfs(i, j)
                cnt += 1

    max_cnt = max(cnt, max_cnt)

print(max_cnt)

File: 백준/Silver/2468. 안전 영역/안전 영역.cc
#include <bits/stdc++.h>
using namespace std;

int n;
int i, j, h, x, y, max_cnt;

bool is_reachable(int x, int y) {
	return (0 <= x && x < n) && (0 <= y && y < n);
}

int arr[101][101];
bool visited[101][101];

const int dxs[4] = {1, 0, -1, 0};
const int dys[4] = {0, 1, 0, -1};

void bfs(int x, int y, int d) {
	queue<pair<int, int>> q;

	q.push({x, y});
	visited[x][y] = true;

	while (q.size()) {
		tie(x, y) = q.front(); q.pop();
		for (i = 0; i < 4; ++i) {
			int nx = x + dxs[i];
			int ny = y + dys[i];

			if (is_reachable(nx, ny) && !visited[nx][ny] && arr[nx][ny] > d) {
				visited[nx][ny] = true;
				q.push({nx, ny});
			}
		}

	}
}

int main() {
	ios_base::sync_with_stdio(false);
	cin.tie(NULL);
	cout.tie(NULL);

	cin >> n;

	for (i = 0; i < n; ++i) {
		for (j = 0; j < n; ++j) {
			cin >> arr[i][j];
		}
	}

	for (h = 0; h <= 100; ++h) {
		fill(&visited[0][0], &visited[0][0] + 101 * 101, 0);

		int cnt = 0;

		for (x = 0; x < n; ++x) {
			for (y = 0; y < n; ++y) {
				if (arr[x][y] > h && !visited[x][y]) {
					bfs(x, y, h);
					cnt++;
				}
			}
		}

		max_cnt = max(max_cnt, cnt);
	}

	cout << max_cnt;

	return 0;
}